src/main/java/rx/internal/operators/OperatorEagerConcatMap.java

```diff
@@ -0,0 +1,299 @@
+        }
+        
+        void cleanup() {
+            List<Subscription> list;
+            synchronized (subscribers) {
+                list = new ArrayList<Subscription>(subscribers);
+                subscribers.clear();
+            }
+            
+            for (Subscription s : list) {
+                s.unsubscribe();
+            }
+        }
+        
+        @Override
+        public void onNext(T t) {
+            Observable<? extends R> observable;
+            
+            try {
+                observable = mapper.call(t);
+            } catch (Throwable e) {
+                Exceptions.throwOrReport(e, actual, t);
+                return;
+            }
+            
+            EagerInnerSubscriber<R> inner = new EagerInnerSubscriber<R>(this, bufferSize);
+            if (cancelled) {
+                return;
+            }
+            synchronized (subscribers) {
+                if (cancelled) {
+                    return;
+                }
+                subscribers.add(inner);
+            }
+            if (cancelled) {
+                return;
+            }
+            observable.unsafeSubscribe(inner);
+            drain();
+        }
+        
+        @Override
+        public void onError(Throwable e) {
+            error = e;
+            done = true;
+            drain();
+        }
+        
+        @Override
+        public void onCompleted() {
+            done = true;
+            drain();
+        }
+        
+        void drain() {
+            if (wip.getAndIncrement() != 0) {
+                return;
+            }
+            int missed = 1;
+            
+            final AtomicLong p = sharedProducer;
+            final Subscriber<? super R> a = actual;
+            
+            for (;;) {
+                
+                if (cancelled) {
+                    cleanup();
+                    return;
+                }
+                
+                EagerInnerSubscriber<R> inner;
+                
+                boolean d = done;
+                synchronized (subscribers) {
+                    inner = subscribers.peek();
+                }
+                boolean empty = inner == null;
+                
+                if (d) {
+                    Throwable err = error;
+                    if (err != null) {
+                        cleanup();
+                        a.onError(err);
+                        return;
+                    } else
+                    if (empty) {
+                        a.onCompleted();
+                        return;
+                    }
+                }
+
+                if (!empty) {
+                    long r = p.get();
+                    long e = 0L;
+                    boolean unbounded = r == Long.MAX_VALUE;
+                    Queue<R> q = inner.queue;
+                    boolean innerDone = false;
+                    
+                    
+                    for (;;) {
+                        d = inner.done;
```

Comment:
There's a bit of reuse of the `d` boolean here for a different meaning (`innerDone`). Could we use a new variable name for this?